fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/OrElseCursor.java

```diff
@@ -66,22 +111,37 @@ public OrElseCursor(@Nonnull RecordCursor<T> inner, @Nonnull Function<Executor,
         if (nextResult != null && !nextResult.hasNext()) {
             return CompletableFuture.completedFuture(nextResult);
         }
-        if (first) {
-            return inner.onNext().thenCompose(result -> {
-                first = false;
-                if (result.hasNext() || result.getNoNextReason().isOutOfBand()) {
-                    // Either have result or do not know whether to select else yet or not.
-                    return CompletableFuture.completedFuture(result);
-                } else {
-                    other = func.apply(getExecutor());
-                    return other.onNext();
-                }
-            }).thenApply(this::postProcess);
-        }
-        if (other != null) {
-            return other.onNext().thenApply(this::postProcess);
+        final CompletableFuture<RecordCursorResult<T>> innerFuture;
+        switch (state) {
+            case USE_INNER:
+                innerFuture = inner.onNext();
+                break;
+            case USE_OTHER:
+                innerFuture = other.onNext();
+                break;
+            case UNDECIDED:
+                innerFuture = inner.onNext().thenCompose(result -> {
+                    if (result.hasNext()) {
+                        // Inner cursor has produced a value, so we take the inner branch.
+                        state = State.USE_INNER;
+                        return CompletableFuture.completedFuture(result);
+                    } else if (result.getNoNextReason().isOutOfBand()) {
```

Comment:
Okay, trying to think about how this interacts with in-band vs out of band limits. I...think it's right that if the child hits an in-band limit (which would only happen if, say, one had a filter over a row limited cursor, and the inner child of that filter stopped hit its limit prior to it no longer returning results) without hitting a value, then it goes down the else path, but I'm not actually sure if that's right. I guess it's consistent with the prior behavior.

Reply:
I believe that what we have here exactly matches what happened before, and I would like to keep it that way.

Reply:
Okay, sounds good.